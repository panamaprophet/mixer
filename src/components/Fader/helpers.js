'use strict';


import {min, max} from 'ramda';


/**
 * @typedef {Object} PointerPosition
 * @property {number} value - absolute value in pixels
 * @property {number} percentage - relative value in percents
 */


/**
 * @returns {number} vertical coordinate from touch or mouse event
 */
export const getY = event => event.touches ? event.touches[0].pageY : event.pageY;

/**
 * @returns {number} horizontal coordinate from touch or move event
 */
export const getX = event => event.touches ? event.touches[0].pageX : event.pageX;

/**
 * @returns {PointerPosition}
 */
export const getPointerVerticalPosition = (position, {top, bottom, height}) => {
    const value = bottom - min(min(position, top), bottom);

    return {
        value,
        percentage: value / (height / 100),
    };
};

/**
 * @returns {PointerPosition}
 */
export const getPointerHorizontalPosition = (position, {width, left}) => {
    const value = max(min(position, left + width), left) - left;

    return {
        value,
        percentage: value / (width / 100),
    }
};