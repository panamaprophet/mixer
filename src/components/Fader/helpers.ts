'use strict';


import {min, max} from 'ramda';

/**
 * @returns {number} vertical coordinate from touch or mouse event
 */
export const getY = event => event.touches ? event.touches[0].pageY : event.pageY;

/**
 * @returns {number} horizontal coordinate from touch or move event
 */
export const getX = event => event.touches ? event.touches[0].pageX : event.pageX;

/**
 * @returns {number} relative position in percents
 */
export const getPointerVerticalPosition = (position, {top, bottom, height}) => {
    let value = position;

    if (value < top) value = top;
    if (value > bottom) value = bottom;

    value = bottom - value;

    return value / (height / 100);
};

/**
 * @returns {number} relative position in percents
 */
export const getPointerHorizontalPosition = (position, {width, left}) => {
    let value = position; // max(min(position, left + width), left) - left;

    const rightBorder = left + width;

    if (value > rightBorder) value = rightBorder;
    if (value < left) value = left;

    value = value - left;

    return value / (width / 100);
};