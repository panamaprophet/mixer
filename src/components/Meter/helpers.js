import {converge, divide, sum, length} from 'ramda';


const COLOR_GREEN = '#05860f';
const COLOR_YELLOW = '#f9ef5c';
const COLOR_RED = '#861615';

/**
 * @param {AudioContext} context
 * @param {Canvas} canvas
 * @returns {CanvasGradient}
 */
export const createMeterGradient = (context, {
    width,
    height,
}) => {
    const gradient = context.createLinearGradient(0, 0, width, height);

    gradient.addColorStop(0.0, COLOR_GREEN);
    gradient.addColorStop(0.8, COLOR_YELLOW);
    gradient.addColorStop(0.9, COLOR_RED);

    return gradient;
}

export const getAverage = converge(divide, [sum, length]);