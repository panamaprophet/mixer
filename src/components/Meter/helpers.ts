import {converge, divide, sum, length} from 'ramda';


const LEVEL_LOW = '#05860f';

const LEVEL_MIDDLE = '#f9ef5c';

const LEVEL_HIGH = '#861615';


/**
 * @param {AudioContext} context
 * @param {Canvas} canvas
 * @returns {CanvasGradient}
 */
export const createMeterGradient = (context: CanvasRenderingContext2D, {
    width,
    height,
}: {
    width: number,
    height: number,
}): CanvasGradient => {
    const gradient = context.createLinearGradient(0, 0, width, height);

    gradient.addColorStop(0.0, LEVEL_LOW);
    gradient.addColorStop(0.8, LEVEL_MIDDLE);
    gradient.addColorStop(0.9, LEVEL_HIGH);

    return gradient;
}

export const getAverage = converge(divide, [sum, length]);