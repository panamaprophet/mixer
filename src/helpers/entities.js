import {
    getNodeParamNormalizedValue,
} from '/helpers/node';

import {
    PLAYBACK_STATUS,
} from '/constants';


const createDelayEntity = ({id, time, feedback, frequency}) => ({
    id,
    name: 'Delay',
    parameters: [{
        name: 'Time',
        id: 'time',
        value: time,
    }, {
        name: 'Feedback',
        id: 'feedback',
        value: feedback,
    }, {
        name: 'Frequency',
        id: 'frequency',
        value: frequency,
    }],
});

const createDistortionEntity = ({id, filterType, frequency, strength}) => ({
    id,
    name: 'Distortion',
    parameters: [{
        name: 'Filter Type',
        id: 'filterType',
        value: filterType,
    }, {
        name: 'Frequency',
        id: 'distFrequency',
        value: frequency,
    }, {
        name: 'Strength',
        id: 'strength',
        value: strength,
    }],
});

export const generateIdByTitle = title => title.replace(/[^A-Za-z]+/gi, '').toLowerCase();

export const createTrackEntity = ({id, title, volume, muted, bypassFX, fx}) => ({
    id,
    title,
    volume,
    isMuted: muted,
    isEffectsDisabled: bypassFX,
    send: {
        delay: getNodeParamNormalizedValue(fx.delay.gain),
        reverb: getNodeParamNormalizedValue(fx.reverb.gain),
        distortion: getNodeParamNormalizedValue(fx.distortion.gain),
    },
});

/**
 * 
 * @param {Effect} effect 
 */
export const createEffectEntity = effect => {
    switch (effect.id) { //@TODO: switch to type
        case 'delay': 
            return createDelayEntity(effect);
        case 'distortion':
            return createDistortionEntity(effect);
        default:
            return null;
    }
};

export const createPlaybackEntity = ({analyser, status = PLAYBACK_STATUS.PAUSED, currentPosition = 0}) => ({
    analyser,
    status,
    currentPosition,
});