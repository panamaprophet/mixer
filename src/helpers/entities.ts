import type {Track, TrackId, TrackState} from '../models/track';
import {Delay, Distortion, Reverb, Send, SendId, SendParamValue} from '/models/sends/index';
import {Playback, PlaybackStatus} from './playback';
import {getNodeParamNormalizedValue} from './node';


export type SendParameterId = string;

export enum SendParameterType {
    RADIO = 'radio',
    FADER = 'fader',
    UNKNOWN = 'unknown',
}

export type SendParameter = {
    id: SendParameterId,
    name: string,
    value: SendParamValue,
    values?: SendParamValue[],
    type: SendParameterType,
}

export type SendEntity = {
    id: SendId,
    name: string,
    parameters: SendParameter[],
}

export type TrackEntity = {
    id: TrackId,
    title: string,
    volume: number,
    state: TrackState,
    isMuted: boolean,
    isSendsEnabled: boolean,
    send: Record<SendId, number>,
}

const FilterTypes: BiquadFilterType[] = ['highpass', 'lowpass', 'bandpass'];


const createReverbEntity = ({id, responses, currentResponse}: Reverb): SendEntity => ({
    id,
    name: 'Reverb',
    parameters: [{
        name: 'Response',
        id: 'currentResponse',
        value: currentResponse,
        values: Object.keys(responses),
        type: SendParameterType.RADIO,
    }],
});

const createDelayEntity = ({id, time, feedback, frequency}: Delay): SendEntity => ({
    id,
    name: 'Delay',
    parameters: [{
        name: 'Time',
        id: 'time',
        value: time,
        type: SendParameterType.FADER,
    }, {
        name: 'Feedback',
        id: 'feedback',
        value: feedback,
        type: SendParameterType.FADER,
    }, {
        name: 'Frequency',
        id: 'frequency',
        value: frequency,
        type: SendParameterType.FADER,
    }],
});

const createDistortionEntity = ({id, filterType, frequency, strength}: Distortion): SendEntity => ({
    id,
    name: 'Distortion',
    parameters: [{
        name: 'Filter Type',
        id: 'filterType',
        value: filterType,
        values: FilterTypes,
        type: SendParameterType.RADIO,
    }, {
        name: 'Frequency',
        id: 'frequency',
        value: frequency,
        type: SendParameterType.FADER,
    }, {
        name: 'Strength',
        id: 'strength',
        value: strength,
        type: SendParameterType.FADER,
    }],
});

export const generateIdByTitle = (title: string): TrackId => title.replace(/[^A-Za-z0-9]+/gi, '').toLowerCase();

export const createTrackEntity = ({id, title, volume, isMuted, isSendsEnabled, sends, state}: Track): TrackEntity => {
    const sendIds: SendId[] = Object.keys(sends);

    const send = sendIds.reduce((result, sendId) => {
        result[sendId] = getNodeParamNormalizedValue(sends[sendId].gain);

        return result;
    }, {});

    return {
        id,
        title,
        volume,
        state,
        isMuted,
        isSendsEnabled,
        send,
    }
}

export const createSendEntity = (send: Send): SendEntity | null => {
    if (send instanceof Delay) {
        return createDelayEntity(send);
    }

    if (send instanceof Reverb) {
        return createReverbEntity(send);
    }

    if (send instanceof Distortion) {
        return createDistortionEntity(send);
    }

    return null;
};

export const createPlaybackEntity = ({
    analyser,
    status = PlaybackStatus.NOT_SET,
    currentPosition = 0,
}: {
    analyser: AnalyserNode,
    status?: PlaybackStatus,
    currentPosition?: number,
}): Playback => ({
    analyser,
    status,
    currentPosition,
});