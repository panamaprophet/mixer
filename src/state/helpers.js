import {
    getNodeParamNormalizedValue,
} from '/helpers/node';


export const STATUS = {
    'PLAYING': 'PLAYING',
    'PAUSED': 'PAUSED',
};


const createTrackEntity = ({id, title, volume, muted, bypassFX, playing, fx}) => ({
    id,
    title,
    volume,
    isMuted: muted,
    isEffectsDisabled: bypassFX,
    isPlaying: playing,
    send: {
        delay: getNodeParamNormalizedValue(fx.delay.gain),
        reverb: getNodeParamNormalizedValue(fx.reverb.gain),
        distortion: getNodeParamNormalizedValue(fx.distortion.gain),
    },
});

const createEffectEntity = ({id, ...rest}) => ({
    id,
    rest,
});


export const createState = mixdesk => {
    const tracks = mixdesk.tracks.map(createTrackEntity);
    const effects = mixdesk.fx.map(createEffectEntity);
    const analyser = mixdesk.analyser;

    const controls = {
        analyser,
        status: STATUS.PAUSED,
    };

    return {
        tracks,
        effects,
        controls,
    };
};