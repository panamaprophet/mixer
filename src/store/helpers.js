import React, {useEffect} from 'react';
import {map, curry, reject, not} from 'ramda';

import {
    TRACK_STATE,
} from '/constants';

import {
    createEffectEntity,
    createPlaybackEntity,
    createTrackEntity,
} from '/helpers/entities';

import {mixdesk} from './mixdesk';


const compact = reject(item => not(Boolean(item)));

const getLoadingState = track => track.loadingState;

const getLoadingStates = ({tracks}) => map(getLoadingState, tracks);

/**
 * Dispatches set ready state action with provided track params
 *
 * @param {function} dispatch
 * @param {Track}
 */
const dispatchSetTrackState = curry((dispatch, {id, state}) => dispatch({
    type: 'TRACK_SET_READY_STATE',
    payload: {
        trackId: id,
        state,
    },
}));

/**
 * Updates tracks statuses in store
 *
 * @param {function} — dispatch
 * @returns {Promise<Track[]>}
 */
export const setTracksStatusesOnLoad = dispatch => 
    Promise
        .all(getLoadingStates(mixdesk))
        .then(map(dispatchSetTrackState(dispatch)));

/**
 * Brings dispatch arguments to console
 *
 * @param {function} — original dispatch
 * @returns {function} — curried version with original dispatch binded
 */
export const getDispatchWithLog = curry((dispatch, args) => {
    console.log('[DISPATCH LOG]', args);

    return dispatch(args);
});

export const createState = mixdesk => {
    const tracks = compact(mixdesk.tracks.map(createTrackEntity));
    const effects = compact(mixdesk.fx.map(createEffectEntity));
    const playback = createPlaybackEntity({
        analyser: mixdesk.analyser,
    });

    return {
        tracks,
        effects,
        playback,
    };
}