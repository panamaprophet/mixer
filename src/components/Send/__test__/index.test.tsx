import React from 'react';
import {shallow} from 'enzyme';
import {Send} from '../';
import {SendParameterType} from '/helpers/entities';


const propsMock = {
    id: 'test',
    name: 'Effect',
    parameters: [{
        id: 'effect-parameter-1',
        name: 'Effect Parameter 1',
        type: SendParameterType.FADER,
        value: 42.0,
    }, {
        id: 'effect-parameter-2',
        name: 'Effect Parameter 2',
        type: SendParameterType.RADIO,
        value: '71.0',
    }],
    onChange: jest.fn(() => jest.fn()),
};


describe('<Send />', () => {
    it('should renders without any errors', () => {
        const wrapper = shallow(<Send {...propsMock} />);

        expect(wrapper).toMatchSnapshot();
    });
});