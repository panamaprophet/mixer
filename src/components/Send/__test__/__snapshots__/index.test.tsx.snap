// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Send /> should renders without any errors 1`] = `
<div
  className="send"
>
  <div
    className="title"
  >
    Effect
  </div>
  <div
    className="parameterValue"
    key="effect-parameter-1"
  >
    <div
      className="parameter"
      key="effect-parameter-1"
    >
      <span
        className="parameterTitle"
      >
        Effect
        :
      </span>
      <div
        className="parameterContainer"
      >
        <Fader
          id="effect-parameter-1"
          name="Effect Parameter 1"
          onChange={[MockFunction]}
          type="fader"
          value={42}
        />
      </div>
    </div>
  </div>
  <div
    className="parameterValue"
    key="effect-parameter-2"
  >
    <div
      className="parameter"
      key="effect-parameter-2"
    >
      <span
        className="parameterTitle"
      >
        Effect
        :
      </span>
      <div
        className="parameterContainer"
      >
        <Radio
          id="effect-parameter-2"
          name="Effect Parameter 2"
          onChange={[MockFunction]}
          type="radio"
          value="71.0"
        />
      </div>
    </div>
  </div>
</div>
`;
